src/history/cursor.rs

```diff
@@ -0,0 +1,608 @@
+        let hist = Box::new(FileBackedHistory::default());
+        (
+            hist,
+            HistoryCursor::new(HistoryNavigationQuery::Normal(LineBuffer::default())),
+        )
+    }
+    fn create_history_at(cap: usize, path: &Path) -> (Box<dyn History>, HistoryCursor) {
+        let hist = Box::new(FileBackedHistory::with_file(cap, path.to_owned()).unwrap());
+        (
+            hist,
+            HistoryCursor::new(HistoryNavigationQuery::Normal(LineBuffer::default())),
+        )
+    }
+
+    #[test]
+    fn accessing_empty_history_returns_nothing() -> Result<()> {
+        let (_hist, cursor) = create_history();
+        assert_eq!(cursor.string_at_cursor(), None);
+        Ok(())
+    }
+
+    #[test]
+    fn going_forward_in_empty_history_does_not_error_out() -> Result<()> {
+        let (hist, mut cursor) = create_history();
+        cursor.forward(&*hist)?;
+        assert_eq!(cursor.string_at_cursor(), None);
+        Ok(())
+    }
+
+    #[test]
+    fn going_backwards_in_empty_history_does_not_error_out() -> Result<()> {
+        let (hist, mut cursor) = create_history();
+        cursor.back(&*hist)?;
+        assert_eq!(cursor.string_at_cursor(), None);
+        Ok(())
+    }
+
+    #[test]
+    fn going_backwards_bottoms_out() -> Result<()> {
+        let (mut hist, mut cursor) = create_history();
+        hist.save(HistoryItem::from_command_line("command1"))?;
+        hist.save(HistoryItem::from_command_line("command2"))?;
+        cursor.back(&*hist)?;
+        cursor.back(&*hist)?;
+        cursor.back(&*hist)?;
+        cursor.back(&*hist)?;
+        cursor.back(&*hist)?;
+        assert_eq!(cursor.string_at_cursor(), Some("command1".to_string()));
+        Ok(())
+    }
+
+    #[test]
+    fn going_forwards_bottoms_out() -> Result<()> {
+        let (mut hist, mut cursor) = create_history();
+        hist.save(HistoryItem::from_command_line("command1"))?;
+        hist.save(HistoryItem::from_command_line("command2"))?;
+        cursor.forward(&*hist)?;
+        cursor.forward(&*hist)?;
+        cursor.forward(&*hist)?;
+        cursor.forward(&*hist)?;
+        cursor.forward(&*hist)?;
+        assert_eq!(cursor.string_at_cursor(), None);
+        Ok(())
+    }
+
+    #[cfg(not(feature = "sqlite"))]
+    #[test]
+    fn appends_only_unique() -> Result<()> {
+        let (mut hist, _) = create_history();
+        hist.save(HistoryItem::from_command_line("unique_old"))?;
+        hist.save(HistoryItem::from_command_line("test"))?;
+        hist.save(HistoryItem::from_command_line("test"))?;
+        hist.save(HistoryItem::from_command_line("unique"))?;
+        assert_eq!(hist.count_all()?, 3);
+        Ok(())
+    }
+
+    #[cfg(not(feature = "sqlite"))]
+    #[test]
+    fn appends_no_empties() -> Result<()> {
+        let (mut hist, _) = create_history();
+        hist.save(HistoryItem::from_command_line(""))?;
+        assert_eq!(hist.count_all()?, 0);
+        Ok(())
+    }
```

Comment:
I think ignoring empty entries is reasonable . I don't think we want to have the db polluted with every mindless enter press.